src/hotspot/share/cds/cdsConfig.cpp

```diff
@@ -527,10 +528,60 @@ bool CDSConfig::current_thread_is_vm_or_dumper() {
   return t != nullptr && (t->is_VM_thread() || t == _dumper_thread);
 }
 
+// If an incompatible VM options is found, return a text message that explains why
+static const char* check_options_incompatible_with_dumping_heap() {
 #if INCLUDE_CDS_JAVA_HEAP
+  if (!UseCompressedClassPointers) {
+    return "UseCompressedClassPointers must be true";
+  }
+
+  // Almost all GCs support heap region dump, except ZGC (so far).
+  if (UseZGC) {
+    return "ZGC is not supported";
+  }
+
+  return nullptr;
+#else
+  return "JVM not configured for writing Java heap objects";
+#endif
+}
+
+void CDSConfig::log_reasons_for_not_dumping_heap() {
+  const char* reason;
+
+  assert(!is_dumping_heap(), "sanity");
+
+  if (_disable_heap_dumping) {
+    reason = "Programmatically disabled";
+  } else {
+    reason = check_options_incompatible_with_dumping_heap();
+  }
+
+  assert(reason != nullptr, "sanity");
+  log_info(cds)("Archived java heap is not supported: %s", reason);
+}
+
+#if INCLUDE_CDS_JAVA_HEAP
+bool CDSConfig::are_vm_options_incompatible_with_dumping_heap() {
+  return check_options_incompatible_with_dumping_heap() != nullptr;
+}
+
+
 bool CDSConfig::is_dumping_heap() {
   // heap dump is not supported in dynamic dump
-  return is_dumping_static_archive() && HeapShared::can_write();
+  if (!is_dumping_static_archive()) {
```

Comment:
I think you can compress these three cases into one if-statement

Reply:
Fixed.